### PR TITLE
Script library management REST API

### PR DESCRIPTION
**Pull request by Wathsara**
This PR Subjected to [Script Library REST API](https://github.com/wso2/product-is/issues/7302)

### CODE REVIEW

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/gen/java/org/wso2/carbon/identity/api/server/script/library/v1/ScriptLibrariesApi.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+import org.wso2.carbon.identity.api.server.script.library.v1.model.Error;
+
+import java.io.File;
+
+import org.wso2.carbon.identity.api.server.script.library.v1.model.ScriptLibraryListResponse;
+import org.wso2.carbon.identity.api.server.script.library.v1.model.ScriptLibraryResponse;
+import org.wso2.carbon.identity.api.server.script.library.v1.ScriptLibrariesApiService;
+
+import javax.validation.Valid;
+import javax.ws.rs.*;
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/gen/java/org/wso2/carbon/identity/api/server/script/library/v1/ScriptLibrariesApi.java]: Better expand the imports with explicit classes.  Wildcard imports are discouraged as it does not show what classes we actually need.

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/gen/java/org/wso2/carbon/identity/api/server/script/library/v1/ScriptLibrariesApiService.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import org.apache.cxf.jaxrs.ext.multipart.Multipart;
+import java.io.InputStream;
+import org.wso2.carbon.identity.api.server.script.library.v1.model.Error;
+import java.io.File;
+import org.wso2.carbon.identity.api.server.script.library.v1.model.ScriptLibraryListResponse;
+import org.wso2.carbon.identity.api.server.script.library.v1.model.ScriptLibraryResponse;
+import javax.ws.rs.core.Response;
+
+
+public interface ScriptLibrariesApiService {
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/gen/java/org/wso2/carbon/identity/api/server/script/library/v1/ScriptLibrariesApiService.java]: We need a class comment explaining what this supposed to do and the purpose, usage

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/gen/java/org/wso2/carbon/identity/api/server/script/library/v1/ScriptLibrariesApiService.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+      public Response addScriptLibrary(String name, InputStream contentInputStream, Attachment contentDetail, String description);
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/gen/java/org/wso2/carbon/identity/api/server/script/library/v1/ScriptLibrariesApiService.java]: Interface does not need "public". any method declared is public by definition

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/gen/java/org/wso2/carbon/identity/api/server/script/library/v1/model/Error.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+import io.swagger.annotations.ApiModelProperty;
+import javax.validation.constraints.*;
+
+
+import io.swagger.annotations.*;
+import java.util.Objects;
+import javax.validation.Valid;
+import javax.xml.bind.annotation.*;
+
+public class Error  {
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/gen/java/org/wso2/carbon/identity/api/server/script/library/v1/model/Error.java]: Class comment is necessary.

File: pom.xml [hunk 5]
```diff
@@ -440,7 +454,7 @@
         <maven.buildnumber.plugin.version>1.4</maven.buildnumber.plugin.version>
         <org.apache.felix.annotations.version>1.2.4</org.apache.felix.annotations.version>
         <identity.governance.version>1.3.15</identity.governance.version>
-        <carbon.identity.framework.version>5.15.36</carbon.identity.framework.version>
+        <carbon.identity.framework.version>5.16.25-SNAPSHOT</carbon.identity.framework.version>
```
Comment (ruwanta) [inline pom.xml]: We do not depend on snapshot versions. It has to be a proper release version.

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/gen/java/org/wso2/carbon/identity/api/server/script/library/v1/ScriptLibrariesApi.java (hunk 1 above)
Comment (Wathsara) [reply components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/gen/java/org/wso2/carbon/identity/api/server/script/library/v1/ScriptLibrariesApi.java]: This class is an autogenerated class from openapi-generator-cxf-wso2

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/gen/java/org/wso2/carbon/identity/api/server/script/library/v1/ScriptLibrariesApiService.java (hunk 2 above)
Comment (Wathsara) [reply components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/gen/java/org/wso2/carbon/identity/api/server/script/library/v1/ScriptLibrariesApiService.java]: This class is an autogenerated class from openapi-generator-cxf-wso2

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/gen/java/org/wso2/carbon/identity/api/server/script/library/v1/ScriptLibrariesApiService.java (hunk 3 above)
Comment (Wathsara) [reply components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/gen/java/org/wso2/carbon/identity/api/server/script/library/v1/ScriptLibrariesApiService.java]: This class is an autogenerated class from openapi-generator-cxf-wso2

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/gen/java/org/wso2/carbon/identity/api/server/script/library/v1/model/Error.java (hunk 4 above)
Comment (Wathsara) [reply components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/gen/java/org/wso2/carbon/identity/api/server/script/library/v1/model/Error.java]: This class is an autogenerated class from openapi-generator-cxf-wso2

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/core/ServerScriptLibrariesService.java [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+
+    /**
+     * Validate the offset.
+     *
+     * @param offset Offset value
+     * @return Validated offset.
+     */
+    private int validateOffset(Integer offset) {
+
+        if (offset != null && offset >= 0) {
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/core/ServerScriptLibrariesService.java]: let's throw an error (Bad request ) if offset is negative. Return default value for null offset

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/core/ServerScriptLibrariesService.java [hunk 7]
```diff
@@ -0,0 +1,439 @@
...
+    /**
+     * Validate the limit.
+     *
+     * @param limit Limit value.
+     * @return Validated limit.
+     */
+    private int validateLimit(Integer limit) {
+
+        final int maximumItemPerPage = IdentityUtil.getMaximumItemPerPage();
+        if (limit != null && limit > 0 && limit <= maximumItemPerPage) {
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/core/ServerScriptLibrariesService.java]: let's throw an error (Bad request ) if limit is negative. Return default value for null limit

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/core/ServerScriptLibrariesService.java [hunk 8]
```diff
@@ -0,0 +1,439 @@
...
+                    Response.Status.INTERNAL_SERVER_ERROR);
+        }
+        if (isScriptLibraryAvailable(scriptLibraryName)) {
+            FunctionLibrary functionLibrary = createScriptLibraryPut(scriptLibraryName, scriptLibraryPUTRequest);
+            try {
+                ScriptLibraryServiceHolder.getScriptLibraryManagementService()
+                        .updateFunctionLibrary(scriptLibraryName, functionLibrary,
+                                ContextLoader.getTenantDomainFromContext());
+            } catch (FunctionLibraryManagementException e) {
+                throw handleScriptLibraryError(e, Constants.ErrorMessage.ERROR_CODE_ERROR_UPDATING_SCRIPT_LIBRARY
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/core/ServerScriptLibrariesService.java]: can't we inline 207 and 208

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/resources/META-INF/cxf/script-libraries-server-v1-cxf.xml [hunk 9]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/resources/META-INF/cxf/script-libraries-server-v1-cxf.xml]: change year

Comment (Wathsara) [reply components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/resources/META-INF/cxf/script-libraries-server-v1-cxf.xml]: [Fixed commit](https://github.com/wso2/identity-api-server/pull/81/commits/f1f24a495063022b48d0fe2a91be8b83c6cf07de)

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/core/ServerScriptLibrariesService.java (hunk 8 above)
Comment (Wathsara) [reply components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/core/ServerScriptLibrariesService.java]: [Fixed commit](https://github.com/wso2/identity-api-server/pull/81/commits/f1f24a495063022b48d0fe2a91be8b83c6cf07de)

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/core/ServerScriptLibrariesService.java (hunk 7 above)
Comment (Wathsara) [reply components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/core/ServerScriptLibrariesService.java]: [Fixed commit](https://github.com/wso2/identity-api-server/pull/81/commits/f1f24a495063022b48d0fe2a91be8b83c6cf07de)

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/core/ServerScriptLibrariesService.java (hunk 6 above)
Comment (Wathsara) [reply components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/core/ServerScriptLibrariesService.java]: [Fixed commit](https://github.com/wso2/identity-api-server/pull/81/commits/f1f24a495063022b48d0fe2a91be8b83c6cf07de)

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.common/src/main/java/org/wso2/carbon/identity/api/server/script/library/common/ScriptLibraryServiceHolder.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (Wathsara) [reply components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.common/src/main/java/org/wso2/carbon/identity/api/server/script/library/common/ScriptLibraryServiceHolder.java]: [Fixed commit](https://github.com/wso2/identity-api-server/pull/81/commits/f1f24a495063022b48d0fe2a91be8b83c6cf07de)

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.common/src/main/java/org/wso2/carbon/identity/api/server/script/library/common/Constants.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (Wathsara) [reply components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.common/src/main/java/org/wso2/carbon/identity/api/server/script/library/common/Constants.java]: [Fixed commit](https://github.com/wso2/identity-api-server/pull/81/commits/f1f24a495063022b48d0fe2a91be8b83c6cf07de)

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/gen/java/org/wso2/carbon/identity/api/server/script/library/v1/ScriptLibrariesApi.java (hunk 1 above)
Comment (ruwanta) [reply components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/gen/java/org/wso2/carbon/identity/api/server/script/library/v1/ScriptLibrariesApi.java]: +1

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/core/ServerScriptLibrariesService.java (hunk 7 above)
Comment (emswbandara) [reply components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/core/ServerScriptLibrariesService.java]: this should be modified as 

if limit is null --> return default (IdentityUtil.getDefaultItemPerPage())
else if limit negative --> throw error
else if limit is greater than maximumItemPerPage, then return maximumItemPerPage
else return the limit

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/core/ServerScriptLibrariesService.java [hunk 12]
```diff
@@ -0,0 +1,438 @@
...
+     * @param scriptLibraryId script library resource ID.
+     */
+    public void deleteScriptLibrary(String scriptLibraryId) {
+
+        if (isScriptLibraryAvailable(scriptLibraryId)) {
+            try {
+                ScriptLibraryServiceHolder.getScriptLibraryManagementService()
+                        .deleteFunctionLibrary(scriptLibraryId, ContextLoader.getTenantDomainFromContext());
+            } catch (FunctionLibraryManagementException e) {
+                throw handleScriptLibraryError(e, Constants.ErrorMessage.ERROR_CODE_ERROR_DELETING_SCRIPT_LIBRARY
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/core/ServerScriptLibrariesService.java]: cant we merge line 252 and 253

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/impl/ScriptLibrariesApiServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+                .build();
+    }
+
+    @Override
+    public Response updateScriptLibrary(String scriptLibraryName, InputStream contentInputStream,
+                                        Attachment contentDetail, String description) {
+
+        ScriptLibraryResponse scriptLibraryResponse =
+                serverScriptLibrariesService.updateScriptLibrary(scriptLibraryName, contentInputStream, description);
+        return Response.ok().entity(scriptLibraryResponse).build();
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/impl/ScriptLibrariesApiServiceImpl.java]: let's avoid sending any response in update requests as discussed in the code-review

Comment (Wathsara) [reply components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/impl/ScriptLibrariesApiServiceImpl.java]: [Fixed  commit](https://github.com/wso2/identity-api-server/pull/81/commits/2800691d849b18ddbdf4fb3acff4c9f75b8e2795)

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/core/ServerScriptLibrariesService.java (hunk 12 above)
Comment (Wathsara) [reply components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/core/ServerScriptLibrariesService.java]: [fixed commit](https://github.com/wso2/identity-api-server/pull/81/commits/2800691d849b18ddbdf4fb3acff4c9f75b8e2795)

File: components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/core/ServerScriptLibrariesService.java (hunk 7 above)
Comment (Wathsara) [reply components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/core/ServerScriptLibrariesService.java]: Negative values are checked from an upper level.

Comment (Wathsara) [reply components/org.wso2.carbon.identity.api.server.script.library/org.wso2.carbon.identity.api.server.script.library.v1/src/main/java/org/wso2/carbon/identity/api/server/script/library/v1/core/ServerScriptLibrariesService.java]: Fixed other validations

**Comment by emswbandara** [comment]
Let's merge the PR once the framework version is updated and travis failure is fixed.

**Pull request merged**